openid-4-verifiable-credential-issuance-1_0.md

```diff
@@ -1119,6 +1121,67 @@ Cache-Control: no-store
 }
 ```
 
+# Callback Endpoint {#callback_endpoint}
+
+This endpoint is used to receive notification from the Wallet whether credential has been successfully received in the Wallet or not. The Credential Issuer needs to return `callback_id` in the Credential Response or a Batch Credential Response for the Wallet to be able to use this Endpoint. Support for this endpoint is OPTIONAL.
+
+This endpoint can be used after the Credential Issuer has sent Credential Response or Batch Credential Response. It enables the Credential Issuer to take subsequent actions after issuance, depending on whether the credential has been accepted and successully stored by the Wallet, rejected by the Wallet, or errors and other unforeseen circumstances have occurred during the Wallet's processing.
+
```

Comment:
I think we may need a bit of extra text to set expectations around how this can go wrong; I suggest something like this:

If the call to this endpoint fails for a temporary reason , the wallet SHOULD retry it, using exponential backoff or other mechanisms to avoid retrying too frequently. In some cases, the issuer will never receive a notification of whether the credential was accepted or not. The call must be idempotent - that is, the issuer must return success if it receives multiple identical calls from the wallet for the same credential.

Reply:
I like the last line, but believe we can simplify it a bit more... how about -

"The Credential Issuer may never receive a callback, even thought a `callback_id` is included in its Credential Response. In this circumstance, it is left to the Credential Issuer to determine how to proceed. 

The Credential Issuer must return success if it receives multiple valid identical calls from the Wallet for the same credential."

Otherwise I fear we might have to start including params specifying how long the callback is valid for, what backoff periods are acceptable and soforth, while IMO that should all be left to the implementer and their desired UX.

Reply:
> while IMO that should all be left to the implementer and their desired UX

I think the problem is this is a UX that is defined by the combination of the issuer behaviour and the wallet behaviour (which may be from different implementers), so neither party acting independently can achieve any particular UX.

If we don't recommend some sort of 'best efforts' in retries from the wallet I start to wonder if the issuer will be unaware of the status of too high a percentage of credentials.

I think if we did go the way you're suggesting it'd at least be worth clarifying:

1. That the callback may not be received for both successful and unsuccessful cases (making it clear that you can't make any assumption that "no callback" => "credential hasn't been stored".
2. That the spec doesn't require the wallet to retry failed requests.

Reply:
I agree it makes sense to clarify the behavior when the callback fails. I think the issuer needs to pre-determine the amount of time that it accepts callbacks for a given `credential_id`. not sure it needs to be a separate sentence/normative requirement. Suggestion:

```suggestion

The callback from the Wallet is idempotent. The Credential Issuer MUST return success if it receives multiple identical calls from the Wallet for the same `callback_id`s.

It is up to the Wallet whether to retry if the call to this endpoint fails. If the Credential Issuer does not receive the callback within the pre-determined amount of time despite providing `callback_id` in the Credential Response, it MUST interpret that as Wallet did not successfully received the credential(s).
```

Reply:
@Sakurann re:

> it MUST interpret that as Wallet did not successfully received the credential(s)

I don't think we should say that. The wallet may have accepted the credential but not been able to notify the issuer for some reason, and we wouldn't want the issuer to try and force the user to obtain a credential that they already have, nor do we want the issuer to revoke a credential that the user thinks they have. I think in this case the issuer has to assume that it doesn't know if the user has the credential or not - so it might want to offer the credential to the user again. It all gets a bit tricky really...

Reply:
played a little with the language. please see the most up to date suggestion in the original comment.

Reply:
I like the edit.